modules/balana-core/src/main/java/org/wso2/balana/cond/xacml3/XACML3HigherOrderFunction.java

```diff
@@ -0,0 +1,364 @@
+        throw new IllegalArgumentException("higher-order functions require use of bags");
+    }
+
+    @Override
+    public EvaluationResult evaluate(List inputs, EvaluationCtx context) {
+
+        Iterator iterator = inputs.iterator();
+
+        // Get the first arg, which is the function.
+        Expression xpr = (Expression) (iterator.next());
+        Function function = null;
+
+        if (xpr instanceof Function) {
+            function = (Function) xpr;
+        }
+
+        // Separate the remaining inputs into primitive data types or bags of primitive types.
+        List<AttributeValue> args = new ArrayList<>();
+        List<BagAttribute> bagArgs = new ArrayList<>();
+
+        while (iterator.hasNext()) {
+            Evaluatable eval = (Evaluatable) (iterator.next());
+            EvaluationResult result = eval.evaluate(context);
+            if (result.indeterminate()) {
+                return result;
+            }
+            if (result.getAttributeValue().returnsBag()) {
+                bagArgs.add((BagAttribute) (result.getAttributeValue()));
+            } else {
+                args.add((AttributeValue) (result.getAttributeValue()));
+            }
+        }
+
+        switch (functionId) {
+            case ID_ANY_OF:
+                return anyAndAllHelper(context, function, args, bagArgs.get(0), false);
+
+            case ID_ALL_OF:
+                return anyAndAllHelper(context, function, args, bagArgs.get(0), true);
+
+            case ID_ANY_OF_ANY:
+                return anyOfAny(context, function, args, bagArgs);
+        }
+        return null;
+    }
+
+    private EvaluationResult anyAndAllHelper(EvaluationCtx context, Function function, List<AttributeValue> values,
+                                             BagAttribute bag, boolean isAllFunction) {
+
+        Iterator it = bag.iterator();
+        while (it.hasNext()) {
+            AttributeValue bagValue = (AttributeValue) (it.next());
+            for (AttributeValue value : values) {
+                EvaluationResult result = getEvaluationResult(context, function, value, bagValue, isAllFunction);
+                if (result != null) {
+                    return result;
+                }
+            }
+        }
+        return new EvaluationResult(BooleanAttribute.getInstance(isAllFunction));
+    }
+
+    private EvaluationResult anyOfAny(EvaluationCtx context, Function function, List<AttributeValue> args,
+                                      List<BagAttribute> bagArgs) {
+
+        // The expression SHALL be evaluated as if the function named in the <Function> argument was applied
+        // between every tuple of the cross product on all bags and the primitive values, and the results were
+        // combined using “urn:oasis:names:tc:xacml:1.0:function:or”
+
+        EvaluationResult result = new EvaluationResult(BooleanAttribute.getInstance(false));
+        if (args.size() > 0) {
```

Comment:
redundant as the for loop already checks this

Reply:
We need this check to have different logic for bags of primitive and primitive